note/note_rfc6962.go

```diff
@@ -0,0 +1,239 @@
+	Timestamp         uint64 `json:"timestamp"`           // The time at which the STH was created
+	SHA256RootHash    []byte `json:"sha256_root_hash"`    // The root hash of the log's Merkle tree
+	TreeHeadSignature []byte `json:"tree_head_signature"` // Log's signature over a TLS-encoded TreeHeadSignature
+	LogID             []byte `json:"log_id"`              // The SHA256 hash of the log's public key
+}
+
+// RFC6962STHToCheckpoint converts the provided RFC6962 JSON representation of a SignedTreeHead structure to
+// a sunlight style signed checkpoint.
+// The passed in verifier must be an RFC6929Verifier containing the correct details for the log which signed the STH.
+func RFC6962STHToCheckpoint(j []byte, v note.Verifier) ([]byte, error) {
+	var sth signedTreeHead
+	if err := json.Unmarshal(j, &sth); err != nil {
+		return nil, err
+	}
+	logName := v.Name()
+	body := fmt.Sprintf("%s\n%d\n%s\n", logName, sth.TreeSize, base64.StdEncoding.EncodeToString(sth.SHA256RootHash))
+	sigBytes := binary.BigEndian.AppendUint32(nil, v.KeyHash())
+	sigBytes = binary.BigEndian.AppendUint64(sigBytes, sth.Timestamp)
+	sigBytes = append(sigBytes, sth.TreeHeadSignature...)
+	sigLine := fmt.Sprintf("\u2014 %s %s", logName, base64.StdEncoding.EncodeToString(sigBytes))
+
+	return []byte(fmt.Sprintf("%s\n%s\n", body, sigLine)), nil
+}
+
+func rfc6962Keyhash(name string, logID [32]byte) uint32 {
+	h := sha256.New()
+	h.Write([]byte(name))
+	h.Write([]byte{0x0A, algRFC6962STH})
+	h.Write(logID[:])
+	r := h.Sum(nil)
+	return binary.BigEndian.Uint32(r)
+}
+
+// rfc6962LogName returns a sunlight checkpoint compatible log name from the
+// passed in CT log root URL.
+//
+// "For example, a log with submission prefix https://rome.ct.example.com/2024h1/ will use rome.ct.example.com/2024h1 as the checkpoint origin line"
+func rfc6962LogName(logURL string) string {
+	logURL = strings.ToLower(logURL)
+	logURL = strings.TrimPrefix(logURL, "http://")
+	logURL = strings.TrimPrefix(logURL, "https://")
+	logURL = strings.TrimSuffix(logURL, "/")
+	return logURL
+}
+
+type rfc6962Verifer struct {
+	name    string
+	keyHash uint32
+	v       func(msg []byte, origin string, sig []byte) bool
+}
+
+// Name returns the name associated with the key this verifier is based on.
+func (v *rfc6962Verifer) Name() string {
+	return v.name
+}
+
+// KeyHash returns a truncated hash of the key this verifier is based on.
+func (v *rfc6962Verifer) KeyHash() uint32 {
+	return v.keyHash
+}
+
+// Verify checks that the provided sig is valid over msg for the key this verifier is based on.
+func (v *rfc6962Verifer) Verify(msg, sig []byte) bool {
+	return v.v(msg, v.name, sig)
+}
+
+func verifyRFC6962(key crypto.PublicKey) func([]byte, string, []byte) bool {
+	return func(msg []byte, origin string, sig []byte) bool {
+		if len(sig) < timestampSize {
+			return false
+		}
+		t := binary.BigEndian.Uint64(sig)
+		sig = sig[timestampSize:]
+		hAlg := sig[0]
+		sAlg := sig[1]
+		sig = sig[2:]
```

Comment:
Did you mean to have sig = sig[2:] twice? I'm sure yes if the tests pass, but what are these two bytes? The firs two are the hash and signature, but why get rid of the two other bytes?